### PR TITLE
UCP/TCP/KEEPALIVE: added processing of auto time

### PR DESCRIPTION
**Pull request by hoopoepg**
- added processing of "auto" timeouts for keepalive
- added parser type UNITS to allow set number with "auto" & "inf"

### CODE REVIEW

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
bot:pipe:retest

File: src/ucp/core/ucp_worker.h [hunk 1]
```diff
@@ -261,6 +268,8 @@ typedef struct ucp_worker {
     ucp_rkey_config_t                rkey_config[UCP_WORKER_MAX_RKEY_CONFIG];
 
     struct {
+        ucs_time_t                   timeout;             /* Time period between keepalive rounds */
```
Comment (brminich) [inline src/ucp/core/ucp_worker.h:271]: why need to keep it on worker if it will be the same for all of them?
Imo, better keep on context

Comment (hoopoepg) [reply src/ucp/core/ucp_worker.h:271]: context may have "auto" value which may be replaced by specific value... replacing inplace is not good idea

Comment (hoopoepg) [reply src/ucp/core/ucp_worker.h:271]: @brminich any other comments?

Comment (brminich) [reply src/ucp/core/ucp_worker.h:271]: no need for replacing inplace. Can just add to the context, not to the config

Comment (hoopoepg) [reply src/ucp/core/ucp_worker.h]: okay

File: src/ucp/core/ucp_context.h [hunk 2]
```diff
@@ -207,6 +207,12 @@ typedef struct ucp_context {
     /* Mask of memory type communication resources */
     uint64_t                      mem_type_access_tls[UCS_MEMORY_TYPE_LAST];
 
+    struct {
+        ucs_time_t                timeout;    /* Time period between keepalive rounds */
+        unsigned                  num_eps;    /* Maximal number of endpoints to check
+                                               * on every round */
+    } keepalive;
```
Comment (brminich) [inline src/ucp/core/ucp_context.h:214]: plz move to non external config, like with other vars

Comment (hoopoepg) [reply src/ucp/core/ucp_context.h]: moved

**Review by brminich — APPROVED** [review]

File: src/ucp/core/ucp_context.c [hunk 3]
```diff
@@ -27,6 +27,16 @@
 
 #define UCP_RSC_CONFIG_ALL    "all"
 
+/* Seconds between keepalive rounds */
+#define UCP_CONTEXT_KEEPALIVE_TIMEOUT 1
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:31]: UCP_CONTEXT_DEFAULT_KEEPALIVE_TIMEOUT

File: src/ucp/core/ucp_context.c [hunk 4]
```diff
@@ -27,6 +27,16 @@
...
+
+/* Maximal number of endpoints to check on every keepalive round */
+#define UCP_CONTEXT_KEEPALIVE_NUM_EPS 128
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:34]: UCP_CONTEXT_KEEPALIVE_DEFAULT_NUM_EPS

File: src/ucp/core/ucp_context.c [hunk 5]
```diff
@@ -27,6 +27,16 @@
...
+
+#define UCP_CONTEXT_AUTO_VALUE(_value, _auto, _auto_val) \
+    (((_value) == (_auto)) ? _auto_val : _value)
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:37]: ( ) around params

**Review by yosefe — COMMENTED** [review]
1. why need KA auto values in both UCP and UCT layers?
2. where do we pass the KA interval from UCP to UCT?

File: src/ucp/core/ucp_context.c (hunk 3 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_context.c]: renamed to UCP_CONTEXT_DEFAULT_KEEPALIVE_TIMEOUT

File: src/ucp/core/ucp_context.c (hunk 4 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_context.c]: renamed to UCP_CONTEXT_KEEPALIVE_DEFAULT_NUM_EPS

File: src/ucp/core/ucp_context.c (hunk 5 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_context.c]: arounded ()

**Review by hoopoepg — COMMENTED** [review]
> 1. why need KA auto values in both UCP and UCT layers?
built-in TCP keepalive engine has separate configuration for now, I just added processing of "auto" value to TCP config

> 2. where do we pass the KA interval from UCP to UCT?
I was going to add it in separate PR - it is internal API change

may we discuss built-in TCP keepalive configuration one more time on Monday? I have a questions...

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

File: src/ucp/core/ucp_context.h [hunk 6]
```diff
@@ -256,6 +256,12 @@ typedef struct ucp_context {
         /* Config environment prefix used to create the context */
         char                      *env_prefix;
 
+        struct {
+            /* Time period between keepalive rounds */
+            ucs_time_t            timeout;
+            /* Maximal number of endpoints to check on every round */
+            unsigned              num_eps;
+        } keepalive;
```
Comment (yosefe) [inline src/ucp/core/ucp_context.h:264]: why needed?

File: src/uct/tcp/tcp_ep.c [hunk 7]
```diff
@@ -428,14 +437,14 @@ static ucs_status_t uct_tcp_ep_keepalive_enable(uct_tcp_ep_t *ep)
...
-        (iface->config.keepalive.cnt == 0) ||
-        (iface->config.keepalive.intvl == 0)) {
+    if (!uct_tcp_keepalive_is_enabled(iface)) {
         return UCS_OK;
     }
 
-    idle_sec  = ucs_max(1, (int)ucs_time_to_sec(iface->config.keepalive.idle));
-    intvl_sec = ucs_max(1, (int)ucs_time_to_sec(iface->config.keepalive.intvl));
+    idle_sec  = uct_tcp_ep_time_seconds(iface->config.keepalive.idle,
+                                        UCT_TCP_EP_KEEPALIVE_IDLE);
```
Comment (yosefe) [inline src/uct/tcp/tcp_ep.c:445]: UCT_TCP_EP_DEFAULT_KEEPALIVE_IDLE

File: src/uct/tcp/tcp_ep.c [hunk 8]
```diff
@@ -428,14 +437,14 @@ static ucs_status_t uct_tcp_ep_keepalive_enable(uct_tcp_ep_t *ep)
...
+    intvl_sec = uct_tcp_ep_time_seconds(iface->config.keepalive.intvl,
+                                        UCT_TCP_EP_KEEPALIVE_INTVL);
```
Comment (yosefe) [inline src/uct/tcp/tcp_ep.c:447]: UCT_TCP_EP_DEFAULT_KEEPALIVE_INTVL

File: src/ucp/core/ucp_context.c [hunk 9]
```diff
@@ -307,9 +307,10 @@ static ucs_config_field_t ucp_config_table[] = {
    "Experimental: enable new protocol selection logic",
    ucs_offsetof(ucp_config_t, ctx.proto_enable), UCS_CONFIG_TYPE_BOOL},
 
-  {"KEEPALIVE_TIMEOUT", "inf",
-   "Time period between keepalive rounds (\"inf\" - disabled).",
-   ucs_offsetof(ucp_config_t, ctx.keepalive_timeout), UCS_CONFIG_TYPE_TIME_UNITS},
+  /* TODO: set for keepalive more reasonable values */
+  {"KEEPALIVE_TIMEOUT", "0",
```
Comment (yosefe) [inline src/ucp/core/ucp_context.c:311]: 1. What is the plan to address this TODO?
2. maybe rename KEEPALIVE_TIMEOUT to KEEPALIVE_INTERVAL in UCP?

Comment (hoopoepg) [reply src/ucp/core/ucp_context.c]: 1. originally we planned to set any pre-defined value after real runs
2. renamed to KEEPALIVE_INTERVAL

File: src/ucp/core/ucp_context.h (hunk 6 above)
Comment (hoopoepg) [reply src/ucp/core/ucp_context.h:264]: used to operate by ucs_time_t values in timeouts

File: src/uct/tcp/tcp_ep.c (hunk 8 above)
Comment (hoopoepg) [reply src/uct/tcp/tcp_ep.c]: renamed to UCT_TCP_EP_DEFAULT_KEEPALIVE_INTVL

File: src/uct/tcp/tcp_ep.c (hunk 7 above)
Comment (hoopoepg) [reply src/uct/tcp/tcp_ep.c]: renamed to UCT_TCP_EP_DEFAULT_KEEPALIVE_IDLE

File: src/ucp/core/ucp_context.h (hunk 6 above)
Comment (yosefe) [reply src/ucp/core/ucp_context.h:264]: i mean, why need num_eps?

Comment (hoopoepg) [reply src/ucp/core/ucp_context.h]: for consistency between config & config.ext
removed

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Review by yosefe — APPROVED** [review]

**Comment by hoopoepg** [comment]
timed out

**Review by yosefe — COMMENTED** [review]
pls squash

**Comment by hoopoepg** [comment]
io_demo not found

**Pull request merged**